clients/src/main/java/org/apache/kafka/clients/consumer/internals/TopicMetadataRequestManager.java

```diff
@@ -0,0 +1,215 @@
+     * @return the future of the metadata request.
+     */
+    public CompletableFuture<Map<String, List<PartitionInfo>>> requestTopicMetadata(final Optional<String> topic) {
+        if (inflightRequests.containsKey(topic)) {
+            return inflightRequests.get(topic).future;
+        }
+
+        TopicMetadataRequestState newRequest = new TopicMetadataRequestState(
+                logContext,
+                topic,
+                retryBackoffMs,
+                retryBackoffMaxMs);
+        inflightRequests.put(topic, newRequest);
+        return newRequest.future;
+    }
+
+    // Visible for testing
+    List<TopicMetadataRequestState> inflightRequests() {
+        return new ArrayList<>(inflightRequests.values());
+    }
+
+    class TopicMetadataRequestState extends RequestState {
+        private final Optional<String> topic;
+        CompletableFuture<Map<String, List<PartitionInfo>>> future;
+
+        public TopicMetadataRequestState(final LogContext logContext,
+                                    final Optional<String> topic,
+                                    final long retryBackoffMs,
+                                    final long retryBackoffMaxMs) {
+            super(logContext, TopicMetadataRequestState.class.getSimpleName(), retryBackoffMs,
+                retryBackoffMaxMs);
+            this.future = new CompletableFuture<>();
+            this.topic = topic;
+        }
+
+        /**
+         * prepare the metadata request and return an
+         * {@link org.apache.kafka.clients.consumer.internals.NetworkClientDelegate.UnsentRequest} if needed.
+         */
+        private Optional<NetworkClientDelegate.UnsentRequest> send(final long currentTimeMs) {
+            if (!this.canSendRequest(currentTimeMs)) {
+                return Optional.empty();
+            }
+            this.onSendAttempt(currentTimeMs);
+
+            final MetadataRequest.Builder request;
+            request = topic.map(t -> new MetadataRequest.Builder(Collections.singletonList(t), allowAutoTopicCreation))
+                .orElseGet(MetadataRequest.Builder::allTopics);
+
+            final NetworkClientDelegate.UnsentRequest unsent = new NetworkClientDelegate.UnsentRequest(
+                request,
+                Optional.empty(),
+                (response, exception) -> {
+                    if (exception != null) {
+                        this.future.completeExceptionally(new KafkaException(exception));
+                        inflightRequests.remove(topic);
+                        return;
+                    }
+
+                    try {
+                        Map<String, List<PartitionInfo>> res = handleTopicMetadataResponse((MetadataResponse) response.responseBody());
+                        future.complete(res);
+                        inflightRequests.remove(topic);
+                    } catch (RetriableException e) {
+                        if (e instanceof TimeoutException) {
```

Comment:
This captures the case when a single request takes long on the broker and times out. However, a separate case is that individual request returns on time but with other RetriableException. In that case, TopicMetadataRequestManager will not clean up inflightRequests even when the client request has timed out.

Reply:
Hey @junrao - Sorry for writing such a long reply - I really need to brush up my memory about the network module as well as the TopicMetadataRequest...

In the [current code](https://github.com/apache/kafka/blob/b49013b73efa25466652d8d8122974e60c927ec4/clients/src/main/java/org/apache/kafka/clients/consumer/internals/TopicMetadataFetcher.java#L99-L149) I think the following is how responses should be handled.

For the case of hard failure that the callback receives a non-null exception:
We should remove the inflight request for all non-retriable and TimeoutException.  We should continue to retry for RetriableException until timer runs out.

For the case of soft failure (the server respond with an error code), it seems like the right thing to do is to continue to retry except for the non-retriable exception.  

It is worth noting that, if the request timeout (API level timeout), I think the callback is completed with TimeoutException hard failure, which should cause the exception being purged from the inflight request.

To your question: if the responses include a retriable exception, I believe the right thing to do is to continue to retry until timeout.

In fact - it seems like `if (e instanceof TimeoutException)` (that I added in the last commit) is actually incorrect.  Because it is a RetriableException at it should be retried per:

```
if (error.exception() instanceof RetriableException)
                            shouldRetry = true;
```

It seems like what's missing here is handling the hard RetriableException and TimeoutException.  I will fix the code and add test cases to cover the scenarios above.